---
title: "TCGA_mRNA_script"
author: "Tara Morrison"
date: "05/11/2018"
output: word_document
---
##Initial set up
Working Directory (Lab) & Libraries
``` {r echo=FALSE}
setwd("~/Desktop/part2_project/Liver_Cancer/TCGA_Counts_mRNA_genelevel")

library(RColorBrewer)
library(gplots)
library(ggplot2)
library(DESeq2)
library(Rtsne)
library(pca3d)
library(limma)
library(genefilter)
library(sva)
library(edgeR)
library(statmod)
```

##TCGA DATA

##Upload pdata (TCGA)
```{r}
pdata=read.table('pdata_rnaseq_genelevel_small.txt', header=T)
rownames(pdata)=pdata$Sample
```

##Set Conditions/Sample Types
There are 3 levels: Primary Tumour, Recurrent Tumour and Normal Tissue
```{r}
conds=(pdata$Sample_Type)
cond_colours=brewer.pal(length(unique(conds)), 'Set1')
names(cond_colours)=unique(conds)
```

##Read in count data
Data taken from TCGA data repository - Liver Cancer mRNA
datatype=HTSeq count data
```{r}
##TCGA DATA COUNTS

ddsHTSeq=DESeqDataSetFromHTSeqCount(sampleTable = pdata, directory= '.', design = ~ Sample_Type)

# These gene IDs have extra numbers at the end - folowing code used to remove
rownames(ddsHTSeq) = gsub(".\\d+$","",rownames(ddsHTSeq),perl=T)

# Load in Gene Names and match up with rownames in ddsHTSeq
gene_names = read.table("gene_names.txt",row.names=1,header = F)
gene_names= gene_names[rownames(ddsHTSeq),]
colnames(gene_names)=c("GeneName","Source","BioType")

colData(ddsHTSeq)$Sample_Type=factor(colData(ddsHTSeq)$Sample_Type, levels=levels(pdata$Sample_Type))

dds=estimateSizeFactors(ddsHTSeq)
dds=estimateDispersions(dds)

```

##Data Normalisation
```{r}
normcounts <- counts(dds, normalized=TRUE)
rawcounts=counts(dds,normalized=FALSE)
log2counts=log2(normcounts+1)

```

##Quality Control
Normalisation is generally ok. Consistent differences between both tumour tissues and normal tissues post-normalisation.
```{r}
quartz()
par(mfrow=c(1,2))

barplot(colSums(counts(dds, normalized=FALSE)), col=cond_colours[pdata$Sample_Type], las=2,cex.names=0.5,main='Pre Normalised Counts')
legend("topleft",levels((conds)),cex=0.5,fill=cond_colours[levels((conds))])

barplot(colSums(counts(dds, normalized=TRUE)), col=cond_colours[pdata$Sample_Type], las=2,cex.names=0.5,main='Post Normalised Counts')
legend('topleft',levels((conds)),cex=0.5,fill=cond_colours[levels((conds))])
```

##Dispersion Plot
```{r echo = FALSE}
dds=nbinomWaldTest(dds)

counts_table=counts(dds, normalized=TRUE)

quartz()
plotDispEsts(dds,main='Dispersion Plot')
```

##Variance Stabilising Transformation
```{r}
vsd <- varianceStabilizingTransformation(dds)
vstMat <- assay(vsd)
vstcounts <- vstMat[order(apply(vstMat,1,sum),decreasing =TRUE),]
```

#PCA Plots & Heatmaps
PCA plots shows good separation between tumour tissues and solid normals. One primary tumour point appears to be anomalous whilst another closer to recurrent tumour tissue type.
Heatmap shows good clustering of solid tissue normals. Heatmap supports PCA plot in that one of the primary tumour points appears to be anomalous and does not closely match other primary tumours whilst another more closely matches recurrent tumour tissues.
```{r echo = FALSE}

#Normal PCA Plot
pca=prcomp(t(assay(vsd)))
quartz()
plot(pca$x, main='PCA Variance Stabilised', pch=21, col='black', bg=cond_colours[pdata$Sample_Type],cex=1)
text(pca$x, as.character(conds), pos=1, cex=0.4)
summary(pca)

#3D PCA Plot with 2D comparison
pca3d(pca, group=conds, col=cond_colours)
quartz()
pca2d(pca, group=conds, col=cond_colours, legend='bottomright')

hmcol = colorRampPalette(brewer.pal(9, 'GnBu'))(100)
quartz()
heatmap.2(cor(vstcounts),trace="none",labRow = pdata$Sample_Type, cexRow = 0.6, labCol = pdata$Sample_Type, cexCol = 0.6, col=hmcol)
```

##Rtsne
tsne shows good separation between all tisuse types. Anomalous primary tumour point is less apparent in the tsne plot. One primary tumour point reamins closely grouped with recurrent tumours.
```{r}
#Normalised counts data into tsne object
set.seed(72)
tsne = Rtsne(t(vstMat), perplexity = 3)
#convert to data frame with 2 dimensions
tsne.df=data.frame(tsne.1=tsne$Y[,1], tsne.2=tsne$Y[,2], Type=as.factor(dds$Sample_Type))
quartz()
ggplot(data = tsne.df, aes(tsne.1,  tsne.2, colour=Type, shape=Type))+
  geom_point(size=4)+
  scale_color_manual(values=unique(cond_colours[tsne.df$Type]))
```

#Statistical Analysis
P threshold is set to 0.05.
log fold change threshold set to 0.7

Comparing Solid_Tissue_Normal vs. Primary_Tumor
Found:  2493  Statistical hits

Comparing Solid_Tissue_Normal vs. Recurrent_Tumor
Found:  3977  Statistical hits

More hits (1484) found for recurrent tumour tissue suggesting that increased mutation lends itself to progression from primary to recurrent.
Good separation of tumour types from solid tissue normal.
Primary tumour shows increased prevalence of down regualtion of gene expression. Recurrent tumours are similar but, comparatively, more genes are also significantly up regulated in recurrent tumours.

Highlighted hits limited to 100.

```{r}
p_threshold=0.05
lfc_threshold=0.7
max_display_hits=100

condition_list=levels(conds)

for (i in 1:(length(condition_list)-1)){
 
  normal=condition_list[3]
  compare=condition_list[i]

  print(paste('Comparing', normal, 'vs.', compare))
  
  res=results(dds, contrast = c('Sample_Type', normal, compare))
  res=res[order(res$padj),]
  
  hits=(rownames(res[((res$padj<=p_threshold) & (abs(res$log2FoldChange)>=lfc_threshold) & (!is.na(res$padj))),]))
  
  if (length(hits) != 0){
    if (length(hits)>max_display_hits){
      hits=hits[1:max_display_hits]
    }
    
  print(paste("Found: ",length(hits)," Statistical hits"))
  
  quartz()
  plot(res$log2FoldChange,-log10(res$padj), ylab='-log10(Adjusted P-value)', xlab='log2 FoldChange', main=paste('Volcano Plot:', normal, 'vs.', compare), pch=19, cex=0.2)
  points(res[hits,'log2FoldChange'],-log(res[hits,'padj'],10),pch=21,cex=0.4,col='turquoise1')
  text(res[hits,]$log2FoldChange,-log10(res[hits,]$padj), labels = gene_names[hits,]$GeneName, pos=3, cex = 0.4)
  abline(h=-log10(0.05), lty=3, col='red')
  abline(v=-1, lty=3, col='red')
  abline(v=1, lty=3, col='red')
 
  quartz()
  heatmap.2(vstMat[hits,],trace='none',col=hmcol,Colv=FALSE,dendrogram='row',main=paste("Heatmap Sig. Hits for", compare),labCol = conds, cexCol = 0.5) 
  } else {
    print(paste("No Results for ",compare))
  }
  
}

```

##Median
```{r echo=FALSE}
median.primary=apply(vstMat[,1:4],1,median)
median.recurrent=apply(vstMat[,5:7],1,median)
median.normal=apply(vstMat[,8:12],1,median)

quartz()
plot(median.normal,median.primary, main="Solid_Tissue_Normal vs. Primary_Tumor", pch=19, col='darkblue', cex=0.4)
points(median.normal[hits],median.primary[hits], pch=21, col='turquoise1')
text(median.normal[hits],median.primary[hits],cex=0.4,pos=4,labels = gene_names[hits,]$GeneName)
abline(a=0,b=1, col='red', lwd=2, cex=0.4, lty=2)

quartz()
plot(median.normal,median.recurrent, main="Solid_Tissue_Normal vs. Recurrent_Tumor", pch=19, col='darkblue', cex=0.4)
points(median.normal[hits],median.recurrent[hits], pch=21, col='turquoise1')
text(median.normal[hits],median.recurrent[hits],cex=0.4,pos=4,labels = gene_names[hits,]$GeneName)
abline(a=0,b=1, col='red', lwd=2, cex=0.4, lty=2)

```











##GTEX DATA

## Load in GTEX Data - WORKING
```{r}

gtex=readRDS("GTEX.rds")

# We obtained a simple table from the raw GTEX sample attributes using:
# cut -f 1,6,7 gtex_sample_attributes.txt  | sort -k2 > gtex_simple_samples.txt
# included: Sample IDs, Tissue, Tissue Description

#All GTEX Samples for all Tissue types
gtex_samples=read.table("gtex_simple_samples.txt",row.names=1,sep="\t",header=F)

#Liver samples only - '-' substituted for '.' to match initial GTEX counts data
liver_ids = rownames(gtex_samples)[gtex_samples$Tissue_Type == "Liver"]
liver_ids = gsub("-",".",liver_ids)
gtex = gtex[,intersect(liver_ids,colnames(gtex))]

gtex[intersect(liver_ids,rownames(gtex)),]

common_ids = intersect(rownames(gtex),rownames(rawcounts))

gtex  = gtex[common_ids, ]
rawcounts = rawcounts[common_ids, ] 
```

#Merging GTEX Data with TCGA Data - WORKING
```{r}
#MERGE GTEX & TCGA MATRICES
mergedcounts=cbind(gtex, rawcounts, deparse.level = 1)

#Labelling columns by Sample Type
colnames(mergedcounts)[1:ncol(gtex)]='GTEX_Normal'
colnames(mergedcounts)[ncol(gtex)+1:ncol(rawcounts)]=paste(pdata$Sample_Type)

#DIentify Conditions and Batches (GTEX vs. TCGA)
conds=as.factor(colnames(mergedcounts))
batch = as.factor(c(rep("GTEX",ncol(gtex)),rep("TCGA",ncol(rawcounts))))
cond_colours = brewer.pal(length(unique(conds)), 'Set1')
batch_colours = c("red","darkblue")

names(cond_colours)=unique(conds)
names(batch_colours)=unique(batch)

```

#OLD CODE
colnames(mergedcounts)[1:20]='GTEX_Normal'
colnames(mergedcounts)[21:32]=paste(pdata$Sample_Type)

#SHORT WAY
conds_merged=as.factor(colnames(mergedcounts))
conds_merged_colours=brewer.pal(length(unique(conds_merged)), 'Set1')
names(conds_merged_colours)=unique(conds_merged)

#LONG WAY
conds_merged2=as.factor(c('GTEX_Normal','GTEX_Normal','GTEX_Normal','GTEX_Normal','GTEX_Normal','GTEX_Normal','GTEX_Normal','GTEX_Normal','GTEX_Normal','GTEX_Normal','GTEX_Normal','GTEX_Normal','GTEX_Normal','GTEX_Normal','GTEX_Normal','GTEX_Normal','GTEX_Normal','GTEX_Normal','GTEX_Normal','GTEX_Normal','Primary_Tumor','Primary_Tumor','Primary_Tumor','Primary_Tumor','Recurrent_Tumor', 'Recurrent_Tumor', 'Recurrent_Tumor', 'Solid_Tissue_Normal','Solid_Tissue_Normal','Solid_Tissue_Normal','Solid_Tissue_Normal', 'Solid_Tissue_Normal'))
 
conds_merged_colours2=brewer.pal(length(unique(conds_merged2)), 'Set1')
names(conds_merged_colours2)=unique(conds_merged2)

mergeddata=read.table('mergeddata.txt', header = T, row.names = 1)



#DESeq Object - Inlcuding Batch Effect
```{r}

#Create colData object fro DESeq
coldata=as.data.frame(conds)
coldata$batch=batch

#coldata$batch=batch
colnames(coldata)=c("Sample_Type", 'Batch')

```

#DESeq Object from merges counts matrix 'mergedcounts' EXCLUDING BATCH (Does not work)
```{r}
mergeddsHTSeq<-DESeqDataSetFromMatrix(countData=mergedcounts,colData=coldata,design=~Sample_Type)

mergedds=estimateSizeFactors(mergeddsHTSeq)
mergedds=estimateDispersions(mergedds)

```

#Data normalisation
```{r}
mergenormcounts <- counts(mergedds, normalized=TRUE)
mergerawcounts=counts(mergedds,normalized=FALSE)
mergelog2counts=log2(mergenormcounts+1)

quartz()
par(mfrow=c(1,2))
barplot(colSums(counts(mergedds, normalized=FALSE)), col=cond_colours[conds], las=2,cex.names=0.5,main='Pre Normalised Counts')
legend("topright",levels(conds),cex=0.5,fill=cond_colours[levels(conds)])

barplot(colSums(counts(mergedds, normalized=TRUE)), col=cond_colours[conds], las=2,cex.names=0.5,main='Post Normalised Counts')
legend("topright",levels(conds),cex=0.5,fill=cond_colours[levels(conds)])

```






NEW APPROACH TAKEN FROM HERE - Decided that expression levels in TCGA are too low to directly compare with GTEX and DESeq is insufficient to normalise. Alternative normalisation appraoches taken

A new matrix (testmatrix) is created excluding coutns that are very low - reduces the number of genes)

#TEST WITH NON ZERO GENE COUNTS
#New fucntion created to test merged coutns matrix for non zero values
```{r}
nonzero <- function(val) {
    return(val>=1)
}

#Test matrix created containing only nonzero gene counts based on above function
testmatrix=mergedcounts[apply(mergedcounts,1,min) > 1,]
```


#DESeq nromalisation - not including batch effects
Basic code used in all appraoches outlined below
```{r}
testHTSeq<-DESeqDataSetFromMatrix(countData=testmatrix,colData=coldata,design=~Sample_Type)

testdds=estimateSizeFactors(testHTSeq)
testdds=estimateDispersions(testdds)

testnormcounts <- counts(testdds, normalized=TRUE)
testrawcounts=counts(testdds,normalized=FALSE)
testlog2counts=log2(testnormcounts+1)

quartz()
par(mfrow=c(1,2))
barplot(colSums(counts(testdds, normalized=FALSE)), col=cond_colours[conds], las=2,cex.names=0.5,main='Pre Normalised Counts')
legend("topright",levels(conds),cex=0.5,fill=cond_colours[levels(conds)])

barplot(colSums(counts(testdds, normalized=TRUE)), col=cond_colours[conds], las=2,cex.names=0.5,main='Post Normalised Counts')
legend("topright",levels(conds),cex=0.5,fill=cond_colours[levels(conds)])

#Plot to check there is ok distribution of gene counts after removing low gene counts
quartz()
plot(testmatrix[,1],testmatrix[,2])
quartz()
boxplot(log2(testmatrix+1))

#Normalisation & Dispersion Plot
dds=nbinomWaldTest(testdds)
test_counts=counts(testdds, normalized=TRUE)

quartz()
plotDispEsts(testdds,main='Dispersion Plot')

#VSD
testvsd <- varianceStabilizingTransformation(testdds)
testvstMat <- assay(testvsd)
testvstcounts <- testvstMat[order(apply(testvstMat,1,sum),decreasing =TRUE),]
testpca=prcomp(t(assay(testvsd)))

```

The normalisation is generally poor using DESeq functions
Decided to look at other normalisation options to see whether we can compare the GTEX TCGA batch effect


#2 - normal DESeq Normalisation and Analysis using test DESeq objects above

```{r}
#VSD
testvsd <- varianceStabilizingTransformation(testdds)
testvstMat <- assay(testvsd)
testvstcounts <- testvstMat[order(apply(testvstMat,1,sum),decreasing =TRUE),]

#PCA Plot
testpca=prcomp(t(assay(testvsd)))

quartz()
plot(testpca$x, main='PCA Variance Stabilised for Merged Data', pch=21, col='black', bg=cond_colours[conds], cex=1)
text(testpca$x, as.character(conds), pos=1, cex=0.4)

summary(testpca)

#Heatmap
quartz()
heatmap.2(cor(test_counts),trace="none",labRow = coldata$Sample_Type, cexRow = 0.6, labCol = coldata$Sample_Type, cexCol = 0.6, col=hmcol)

#Rtsne
set.seed(72)
testtsne = Rtsne(t(testvstMat), perplexity = 3, check_duplicates=F)
testtsne.df=data.frame(testtsne.1=testtsne$Y[,1], testtsne.2=testtsne$Y[,2], Type=as.factor(testdds$Sample_Type))

quartz()
ggplot(data = testtsne.df, aes(testtsne.1, testtsne.2, colour=Type, shape=Type))+
  geom_point(size=4)+
  scale_color_manual(values=unique(cond_colours[testtsne.df$Type]))
```


#APPROACH 3 - limma batch correction for data
This attempts to remove Batch effects between GTEX and TCGA
```{r}

#Limma function removal of batch effect
assay(testvsd)=limma::removeBatchEffect(assay(testvsd), testvsd$batch)
quartz()
plotPCA(testvsd, 'batch')

assay(testvsd)=limma::removeBatchEffect(assay(testvsd), testvsd$Sample_Type)
quartz()
plotPCA(testvsd, 'Sample_Type')


testvsd <- varianceStabilizingTransformation(testdds)
testvstMat <- assay(testvsd)
testvstcounts <- testvstMat[order(apply(testvstMat,1,sum),decreasing =TRUE),]

filter1=function(x)(IQR(x)>0.5)

filteredcounts=testvstcounts[genefilter(testvstcounts,filter1),]

pca=prcomp(t(filteredcounts))
plot(pca)

quartz()
plot(pca$x, main='PCA Variance Stabilised for Merged Data', pch=21, col='black', bg=cond_colours[conds], cex=1)
text(pca$x, as.character(conds), pos=1, cex=0.4)

```

#APPROACH 4 - TMM
```{r}

#Type=factor(substring(colnames(mergedcounts),1))
#Time=factor(substring(colnames(mergedcounts),1,))

dge=DGEList(counts=testmatrix)

dge$samples$Type=conds
dge$samples$Batch=batch

dge=calcNormFactors(dge, method = 'TMM')

quartz()
boxplot(log2(dge$counts+1), las=2, col=cond_colours[conds], main='log2 counts', cex.axis=0.6)

quartz()
plotMDS(dge, pch = 19, col= cond_colours[conds], main='Multi Dimensional Scaling (MDS) Plot')
legend('bottomright', legend=levels(conds), col=cond_colours[unique(conds)], pch = 20)


#Doesn't currrently include batch - when batch is added omits GTEX batch
design=model.matrix(~0+batch, data = dge$samples)
rownames(design)=colnames(dge)

dge=estimateDisp(dge, design, robust=T)
dge$common.dispersion
quartz()
plotBCV(dge)

dge=estimateGLMCommonDisp(dge)
dge=estimateTagwiseDisp(dge)
fit = glmFit(dge, design, robust = T)
quartz()

fit2=glmQLFit(dge, design)
#Compare GTEX normal to Primary Tumour (2), Recurrent Tumour (3), Solid 
qlf=glmQLFTest(fit2, coef = 2)
qlf=glmQLFTest(fit2, coef = 3)
qlf=glmQLFTest(fit2, coef = 4)
```


#Alternative Normalisation? - SVA

```{r}
mod=model.matrix(~as.factor(conds), data = testmatrix)
mod0=model.matrix(~as.factor(batch), data = testmatrix)
sva=sva(test_counts, mod, mod0)

#Adjusting for Surrogate Variables
pValues=f.pvalue(test_counts,mod,mod0)
qValues=p.adjust(pValues, method='BH')

modSv = cbind(mod,sva$sv)
mod0Sv = cbind(mod0,sva$sv)
pValuesSv = f.pvalue(test_counts,modSv,mod0Sv)
qValuesSv = p.adjust(pValuesSv,method="BH")

#ComBat Function to adjust for batch effect between GTEX and TCGA - TO RETURN TO
batch
modcombat=model.matrix(~1, data = testmatrix)
combat_test=ComBat(dat = test_counts, batch = batch, mod=modcombat,par.prior=TRUE, prior.plots=FALSE)

#Simplified ComBat approach - TO RETURN TO 
modBatch=model.matrix(~as.factor(conds) + as.factor(batch), data = testmatrix)
mod0Batch=model.matrix(~as.factor(batch), data = testmatrix)

pValuesBatch = f.pvalue(test_counts,modBatch,mod0Batch)
qValuesBatch = p.adjust(pValuesBatch,method="BH")

#sva for sequencing
mod1=model.matrix(~conds)
modnull=cbind(mod1[,1])
svseq=svaseq(test_counts, mod1, modnull)$sv

plot(svseq, pch=19, col=cond_colours[conds])
```







#PCA Plots & Heatmaps

```{r}
#Prcomp PCA
mergepca=prcomp(t(assay(mergevsd)))
quartz()
plot(mergepca$x, main='PCA Variance Stabilised for Merged Data', pch=21, col='black', bg=cond_colours[conds], cex=1)
text(mergepca$x, as.character(conds), pos=1, cex=0.4)

summary(mergepca)

#Princomp PCA
quartz()
mergepca2=princomp(assay(mergevsd))
plot(mergepca2$loadings, main='PCA Variance Stabilised for Merged Data', pch=21, col='black', bg=conds_merged_colours[coldata$Sample_Type], cex=1)
text(mergepca2$loadings, as.character(conds_merged2), pos=1, cex=0.4)

summary(mergepca2)

quartz()
heatmap.2(cor(mergevstcounts),trace="none",labRow = coldata$Sample_Type, cexRow = 0.6, labCol = coldata$Sample_Type, cexCol = 0.6, col=hmcol)

set.seed(72)
mergetsne = Rtsne(t(mergevstMat), perplexity = 3)
mergetsne.df=data.frame(mergetsne.1=mergetsne$Y[,1], mergetsne.2=mergetsne$Y[,2], Type=as.factor(mergedds$Sample_Type))

quartz()
ggplot(data = mergetsne.df, aes(mergetsne.1,  mergetsne.2, colour=Type, shape=Type))+
  geom_point(size=4)+
  scale_color_manual(values=unique(conds_merged_colours2[mergetsne.df$Type]))
