---
title: "TCGA_mRNA_script"
author: "Tara Morrison"
date: "05/11/2018"
output: word_document
---
##Initial set up
Working Directory (Lab) & Libraries
``` {r echo=FALSE, r eval=F}
setwd("~/Desktop/part2_project/Liver_Cancer/TCGA_Counts_mRNA_genelevel")

library(RColorBrewer)
library(gplots)
library(ggplot2)
library(DESeq2)
library(Rtsne)
library(pca3d)
library(limma)
library(genefilter)
library(sva)
library(edgeR)
library(statmod)
```

##TCGA DATA

##Upload pdata (TCGA)
```{r}
pdata=read.table('pdata_rnaseq_genelevel_small.txt', header=T)
rownames(pdata)=pdata$Sample
```

##Set Conditions/Sample Types
There are 3 levels: Primary Tumour, Recurrent Tumour and Normal Tissue
```{r}
conds=(pdata$Sample_Type)
cond_colours=brewer.pal(length(unique(conds)), 'Set1')
names(cond_colours)=unique(conds)
```

##Read in count data
Data taken from TCGA data repository - Liver Cancer mRNA
datatype=HTSeq count data
```{r}
##TCGA DATA COUNTS

ddsHTSeq=DESeqDataSetFromHTSeqCount(sampleTable = pdata, directory= '.', design = ~ Sample_Type)

# These gene IDs have extra numbers at the end - folowing code used to remove
rownames(ddsHTSeq) = gsub(".\\d+$","",rownames(ddsHTSeq),perl=T)

# Load in Gene Names and match up with rownames in ddsHTSeq
gene_names = read.table("gene_names.txt",row.names=1,header = F)
gene_names= gene_names[rownames(ddsHTSeq),]
colnames(gene_names)=c("GeneName","Source","BioType")

#colData(ddsHTSeq)$Sample_Type=factor(colData(ddsHTSeq)$Sample_Type, levels=levels(pdata$Sample_Type))

```

##Data Normalisation
```{r}
normcounts <- counts(dds, normalized=TRUE)
rawcounts=counts(dds,normalized=FALSE)
log2counts=log2(normcounts+1)

```

##Quality Control
Normalisation is generally ok. Consistent differences between both tumour tissues and normal tissues post-normalisation.
```{r}
quartz()
par(mfrow=c(1,2))

barplot(colSums(counts(dds, normalized=FALSE)), col=cond_colours[conds], las=2,cex.names=0.5,main='Pre Normalised Counts', names.arg =conds)
legend("topleft",levels(conds),cex=0.5,fill=cond_colours)

barplot(colSums(counts(dds, normalized=TRUE)), col=cond_colours[conds], las=2,cex.names=0.5,main='Post Normalised Counts', names.arg =conds)
legend("topleft",levels(conds),cex=0.5,fill=cond_colours)
```

##Dispersion Plot
```{r echo = FALSE}
dds=nbinomWaldTest(dds)

counts_table=counts(dds, normalized=TRUE)

quartz()
plotDispEsts(dds,main='Dispersion Plot')
```

##Variance Stabilising Transformation
```{r}
vsd <- varianceStabilizingTransformation(dds)
vstMat <- assay(vsd)
vstcounts <- vstMat[order(apply(vstMat,1,sum),decreasing =TRUE),]
```

#PCA Plots & Heatmaps
PCA plots shows good separation between tumour tissues and solid normals. One primary tumour point appears to be anomalous whilst another closer to recurrent tumour tissue type.
Heatmap shows good clustering of solid tissue normals. Heatmap supports PCA plot in that one of the primary tumour points appears to be anomalous and does not closely match other primary tumours whilst another more closely matches recurrent tumour tissues.
```{r echo = FALSE}

#Normal PCA Plot
pca=prcomp(t(assay(vsd)))
quartz()
plot(pca$x, main='PCA Variance Stabilised', pch=21, col='black', bg=cond_colours[conds],cex=1)
text(pca$x, as.character(conds), pos=1, cex=0.4)
summary(pca)

#3D PCA Plot with 2D comparison
pca3d(pca, group=conds, col=cond_colours, legend = 'bottomright')
quartz()
pca2d(pca, group=conds, col=cond_colours, legend='bottomright')

hmcol = colorRampPalette(brewer.pal(9, 'GnBu'))(100)
quartz()
heatmap.2(cor(vstcounts),trace="none",labRow = conds, cexRow = 0.6, labCol = pdata$Sample_Type, cexCol = 0.6, col=hmcol, main = 'Heatmap of Correlation of the VST Counts')
```

##Rtsne
tsne shows good separation between all tisuse types. Anomalous primary tumour point is less apparent in the tsne plot. One primary tumour point reamins closely grouped with recurrent tumours.
```{r}
#Normalised counts data into tsne object
set.seed(72)
tsne = Rtsne(t(vstMat), perplexity = 3)
#convert to data frame with 2 dimensions
tsne.df=data.frame(tsne.1=tsne$Y[,1], tsne.2=tsne$Y[,2], Type=as.factor(dds$Sample_Type))
quartz()
ggplot(data = tsne.df, aes(tsne.1,  tsne.2, colour=Type, shape=Type))+
  geom_point(size=4)+
  scale_color_manual(values=unique(cond_colours[tsne.df$Type]))
```

#Statistical Analysis
P threshold is set to 0.05.
log fold change threshold set to 0.7

Comparing Solid_Tissue_Normal vs. Primary_Tumor
Found:  2493  Statistical hits

Comparing Solid_Tissue_Normal vs. Recurrent_Tumor
Found:  3977  Statistical hits

More hits (1484) found for recurrent tumour tissue suggesting that increased mutation lends itself to progression from primary to recurrent.
Good separation of tumour types from solid tissue normal.
Primary tumour shows increased prevalence of down regualtion of gene expression. Recurrent tumours are similar but, comparatively, more genes are also significantly up regulated in recurrent tumours.

Highlighted hits limited to 100.

```{r}
p_threshold=0.05
lfc_threshold=0.7
max_display_hits=100

condition_list=levels(conds)

for (i in 2:(length(condition_list))){
 
  normal=condition_list[1]
  compare=condition_list[i]

  print(paste('Comparing', normal, 'vs.', compare))
  
  res=results(dds, contrast = c('Sample_Type', normal, compare))
  res=res[order(res$padj),]
  
  hits=(rownames(res[((res$padj<=p_threshold) & (abs(res$log2FoldChange)>=lfc_threshold) & (!is.na(res$padj))),]))

   print(paste("Found: ",length(hits)," Statistical hits"))
   
  if (length(hits) != 0){
    if (length(hits)>max_display_hits){
      hits=hits[1:max_display_hits]
    }
    
  quartz()
  plot(res$log2FoldChange,-log10(res$padj), ylab='-log10(Adjusted P-value)', xlab='log2 FoldChange', main=paste('Volcano Plot:', normal, 'vs.', compare), pch=19, cex=0.2)
  points(res[hits,'log2FoldChange'],-log(res[hits,'padj'],10),pch=21,cex=0.4,col='turquoise1')
  text(res[hits,]$log2FoldChange,-log10(res[hits,]$padj), labels = gene_names[hits,]$GeneName, pos=3, cex = 0.4)
  abline(h=-log10(0.05), lty=3, col='red')
  abline(v=-1, lty=3, col='red')
  abline(v=1, lty=3, col='red')
 
  quartz()
  heatmap.2(vstMat[hits,],trace='none',col=hmcol,Colv=FALSE,dendrogram='row',main=paste("Heatmap Sig. Hits for", compare),labCol = conds, cexCol = 0.5) 
  } else {
    print(paste("No Results for ",compare))
  }
  
}

```

##Median
```{r echo=FALSE}
median.primary=apply(vstMat[,1:4],1,median)
median.recurrent=apply(vstMat[,5:7],1,median)
median.normal=apply(vstMat[,8:12],1,median)

quartz()
plot(median.normal,median.primary, main="Normal vs. Primary_Tumor", pch=19, col='darkblue', cex=0.4)
points(median.normal[hits],median.primary[hits], pch=21, col='turquoise1')
text(median.normal[hits],median.primary[hits],cex=0.4,pos=4,labels = gene_names[hits,]$GeneName)
abline(a=0,b=1, col='red', lwd=2, cex=0.4, lty=2)

quartz()
plot(median.normal,median.recurrent, main="Normal vs. Recurrent_Tumor", pch=19, col='darkblue', cex=0.4)
points(median.normal[hits],median.recurrent[hits], pch=21, col='turquoise1')
text(median.normal[hits],median.recurrent[hits],cex=0.4,pos=4,labels = gene_names[hits,]$GeneName)
abline(a=0,b=1, col='red', lwd=2, cex=0.4, lty=2)

```


##Save Plots in PDF for TCGA Analysis
```{r}
par(mfrow=c(1,2))

barplot(colSums(counts(dds, normalized=FALSE)), col=cond_colours[conds], las=2,cex.names=0.5,main='Pre Normalised Counts', names.arg =conds)
legend("topleft",levels(conds),cex=0.5,fill=cond_colours)

barplot(colSums(counts(dds, normalized=TRUE)), col=cond_colours[conds], las=2,cex.names=0.5,main='Post Normalised Counts', names.arg =conds)
legend("topleft",levels(conds),cex=0.5,fill=cond_colours)

plotDispEsts(dds,main='Dispersion Plot')

plot(pca$x, main='PCA Variance Stabilised', pch=21, col='black', bg=cond_colours[conds],cex=1)
text(pca$x, as.character(conds), pos=1, cex=0.4)

pca3d(pca, group=conds, col=cond_colours, legend = 'bottomright')
pca2d(pca, group=conds, col=cond_colours, legend='bottomright')

heatmap.2(cor(vstcounts),trace="none",labRow = conds, cexRow = 0.6, labCol = pdata$Sample_Type, cexCol = 0.6, col=hmcol, main = 'Heatmap of Correlation of the VST Counts')

```






##GTEX DATA

## Load in GTEX Data
```{r}

gtex=readRDS('GTEX.rds')

# We obtained a simple table from the raw GTEX sample attributes using:
# cut -f 1,6,7 gtex_sample_attributes.txt  | sort -k2 > gtex_simple_samples.txt
# included: Sample IDs, Tissue, Tissue Description

#All GTEX Samples for all Tissue types
gtex_samples=read.table('gtex_simple_samples.txt',row.names=1,sep="\t",header=F)
colnames(gtex_samples)=c('Tissue_Type', 'Description')

#Liver samples only - '-' substituted for '.' to match initial GTEX counts data
liver_ids = rownames(gtex_samples)[gtex_samples$Tissue_Type == "Liver"]
liver_ids = gsub("-",".",liver_ids)

gtex = gtex[,intersect(liver_ids,colnames(gtex))]

#Need to remove decimal in gtex gene names after initial input
#rownames(gtex)=gsub(".\\d+$","",rownames(gtex),perl=T)

common_ids = intersect(rownames(gtex),rownames(rawcounts))

gtex  = gtex[common_ids, ]
gtex = gtex[,1:20]
rawcounts = rawcounts[common_ids, ] 
```

#Merging GTEX Data with TCGA Data
```{r}
#MERGE GTEX & TCGA MATRICES
mergedcounts=cbind(gtex, rawcounts, deparse.level=1)

#Labelling columns by Sample Type
colnames(mergedcounts)[1:ncol(gtex)]='Normal'
colnames(mergedcounts)[ncol(gtex)+1:ncol(rawcounts)]=paste(pdata$Sample_Type)

#Identify Conditions and Batches (GTEX vs. TCGA)
conds=as.factor(colnames(mergedcounts))
cond_colours = brewer.pal(length(unique(conds)), 'Set1')
names(cond_colours)=unique(conds)

batch = as.factor(c(rep("GTEX",ncol(gtex)),rep("TCGA",ncol(rawcounts))))
batch_colours = c("red","darkblue")
names(batch_colours)=unique(batch)

```


#ColData object produced - Inlcuding Batch Effect
```{r}

#Create colData object fro DESeq
coldata=as.data.frame(conds)
coldata$batch=batch

colnames(coldata)=c("Sample_Type", 'Batch')

```

#DESeq Object from merged counts matrix 'mergedcounts' EXCLUDING BATCH
```{r}
mergeddsHTSeq<-DESeqDataSetFromMatrix(countData=mergedcounts,colData=coldata,design=~Sample_Type+Batch)

mergedds=estimateSizeFactors(mergeddsHTSeq)
mergedds=estimateDispersions(mergedds)

```

#Data normalisation - Using merged data
```{r}
mergenormcounts <- counts(mergedds, normalized=TRUE)
mergerawcounts=counts(mergedds,normalized=FALSE)
mergelog2counts=log2(mergenormcounts+1)

quartz()
par(mfrow=c(1,2))
barplot(colSums(counts(mergedds, normalized=FALSE)), col=cond_colours[conds], las=2,cex.names=0.5,main='Pre Normalised Counts')
legend("topright",levels(conds),cex=0.5,fill=cond_colours[levels(conds)])

barplot(colSums(counts(mergedds, normalized=TRUE)), col=cond_colours[conds], las=2,cex.names=0.5,main='Post Normalised Counts')
legend("topright",levels(conds),cex=0.5,fill=cond_colours[levels(conds)])

```


#Disperison Pot - merged data
```{r}
mergedds=nbinomWaldTest(mergedds)
counts_table=counts(mergedds, normalized=TRUE)

quartz()
plotDispEsts(mergedds,main='Dispersion Plot')
```


#VST - merged data
```{r}
#blind added - to ignore variation due to batch effect
#If many of genes have large differences in counts due to the experimental design, it is important to set blind=FALSE for downstream analysis.

mergevsd <- varianceStabilizingTransformation(mergedds, blind = F)
mergevstMat <- assay(mergevsd)
mergevstcounts <- mergevstMat[order(apply(mergevstMat,1,sum),decreasing =TRUE),]

```


#PCA Plots & Heatmaps - merged data
```{r}
#Prcomp PCA
pca=prcomp(t(assay(mergevsd)))
quartz()
plot(pca$x, main='PCA Variance Stabilised for Merged Data', pch=c(21,24)[batch], col='black', bg=cond_colours[conds], cex=1)
text(pca$x, as.character(conds), pos=1, cex=0.4)

summary(pca)

#Princomp PCA
pca2=princomp(assay(vsd))
quartz()
plot(pca2$loadings, main='PCA Variance Stabilised for Merged Data', pch=21, col='black', bg=cond_colours[conds], cex=1)
text(pca2$loadings, as.character(conds), pos=1, cex=0.4)

summary(pca2)

quartz()
heatmap.2(cor(mergevstcounts),trace="none",labRow = coldata$Sample_Type, cexRow = 0.6, labCol = coldata$Sample_Type, cexCol = 0.6, col=hmcol)

set.seed(72)
tsne = Rtsne(t(mergevstMat), perplexity = 3)
tsne.df=data.frame(tsne.1=tsne$Y[,1], tsne.2=tsne$Y[,2], Type=as.factor(mergedds$Sample_Type), Batch=as.factor(mergedds$Batch))

quartz()
ggplot(data = tsne.df, aes(tsne.1,  tsne.2, colour=Type, shape=Batch))+
  geom_point(size=4)+
  scale_color_manual(values=unique(cond_colours[tsne.df$Type]))
```


#Statistical Analysis - Merged Data
(GTEX/TCGA Merged) Comparing Normal vs. Primary_Tumor
Found:  3902  Statistical hits

(GTEX/TCGA Merged) Comparing Normal vs. Recurrent_Tumor
Found:  5630  Statistical hits
```{r}
p_threshold=0.05
lfc_threshold=0.7
max_display_hits=100

condition_list=levels(conds)

for (i in 2:(length(condition_list))){
 
  normal=condition_list[1]
  compare=condition_list[i]

  print(paste('(GTEX/TCGA Merged) Comparing', normal, 'vs.', compare))
  
  res=results(mergedds, contrast = c('Sample_Type', normal, compare))
  res=res[order(res$padj),]
  
  hits=(rownames(res[((res$padj<=p_threshold) & (abs(res$log2FoldChange)>=lfc_threshold) & (!is.na(res$padj))),]))
  
   print(paste("Found: ",length(hits)," Statistical hits"))
   
  if (length(hits) != 0){
    if (length(hits)>max_display_hits){
      hits=hits[1:max_display_hits]
    }
    
  quartz()
  plot(res$log2FoldChange,-log10(res$padj), ylab='-log10(Adjusted P-value)', xlab='log2 FoldChange', main=paste('Volcano Plot:', normal, 'vs.', compare,'(GTEX/TCGA Merged)'), pch=19, cex=0.2)
  points(res[hits,'log2FoldChange'],-log(res[hits,'padj'],10),pch=21,cex=0.4,col='turquoise1')
  text(res[hits,]$log2FoldChange,-log10(res[hits,]$padj), labels = gene_names[hits,]$GeneName, pos=3, cex = 0.4)
  abline(h=-log10(0.05), lty=3, col='red')
  abline(v=-1, lty=3, col='red')
  abline(v=1, lty=3, col='red')
 
  quartz()
  heatmap.2(mergevstMat[hits,],trace='none',col=hmcol,Colv=FALSE,dendrogram='row',main=paste("Heatmap Sig. Hits for", compare),labCol = conds, cexCol = 0.5) 
  } else {
    print(paste("No Results for ",compare))
  }
  
}
```


#Medians
```{r}

median.primary=apply(mergevstMat[,21:24],1,median)
median.recurrent=apply(mergevstMat[,25:27],1,median)
median.normal=apply(mergevstMat[,c(1:20, 28:32)], 1, median)

quartz()
plot(median.normal,median.primary, main="Normal vs. Primary_Tumor (GTEX/TCGA Merged)", pch=19, col='darkblue', cex=0.4)
points(median.normal[hits],median.primary[hits], pch=21, col='turquoise1')
text(median.normal[hits],median.primary[hits],cex=0.4,pos=4,labels = gene_names[hits,]$GeneName)
abline(a=0,b=1, col='red', lwd=2, cex=0.4, lty=2)

quartz()
plot(median.normal,median.recurrent, main="Normal vs. Recurrent_Tumor (GTEX/TCGA Merged)", pch=19, col='darkblue', cex=0.4)
points(median.normal[hits],median.recurrent[hits], pch=21, col='turquoise1')
text(median.normal[hits],median.recurrent[hits],cex=0.4,pos=4,labels = gene_names[hits,]$GeneName)
abline(a=0,b=1, col='red', lwd=2, cex=0.4, lty=2)

```


##Voom Limma Approach - Using Si's code

#Set-Up
```{r}

dge=DGEList(counts=mergedcounts)

dge$samples$group=conds
dge$samples$batch=batch

cpm=cpm(dge)
lcpm=cpm(dge, log=T)

keep.exprs=rowSums(cpm>1)>=3
dge=dge[keep.exprs,,keep.lib.sizes=F]
dim(dge)

dge=calcNormFactors(dge, method='TMM')
dge$samples$norm.factors

quartz()
boxplot(log2(dge$counts+1), las=2, col=cond_colours[conds], main="Pre Normalised Data", ylab="Log-cpm")

quartz()
plotMDS(dge, pch = 19, col= cond_colours[conds], main='Multi Dimensional Scaling (MDS) Plot')
legend('bottomright', legend=levels(conds), col=cond_colours[unique(conds)], pch = 20)

lcpm=cpm(dge, log=T)
quartz()
boxplot(lcpm, las=2, col=cond_colours[conds], main="Normalised Data",ylab="Log-cpm")
```


#Design only include conditions
```{r}
design=model.matrix(~0+conds)
contr.matrix=makeContrasts(condsPrimary_Tumor-condsNormal, condsRecurrent_Tumor-condsNormal, levels = colnames(design))

quartz()
 v=voom(dge, design, plot = T)

vfit=lmFit(v, design)
vfit=contrasts.fit(vfit, contrasts = contr.matrix)
efit=eBayes(vfit)
contrast1=efit

pca=prcomp(t(v$E))
quartz()
plot(pca$x, pch=c(21,24)[batch], bg=cond_colours[conds], cex=1, title(main = 'PCA Plot'))
  text(pca$x,as.character(coldata$Batch),cex=0.3,pos=1)
  text(pca$x,as.character(colnames(v$E)),cex=0.3,pos = 3)

quartz()
plotSA(efit, main='Mean-Variance Trend')

quartz()
plotMDS(efit, pch = 19, col= cond_colours[conds], main='Multi Dimensional Scaling (MDS) Plot')
  legend('bottomright', legend=levels(conds), col=cond_colours[unique(conds)], pch = 20)

summary(decideTests(efit))

quartz()
volcanoplot(efit, coef=1,names=as.character(gene_names[rownames(efit),1]), highlight = 50, main='Primary Tumour vs. Normal')
  abline(h=-log10(0.05), lty=3, col='red')
  abline(v=-1, lty=3, col='red')
  abline(v=1, lty=3, col='red')
  
quartz()
volcanoplot(efit, coef=2, names=as.character(gene_names[rownames(efit),1]), highlight = 50, main='Recurrent Tumour vs. Normal') 
  abline(h=-log10(0.05), lty=3, col='red')
  abline(v=-1, lty=3, col='red')
  abline(v=1, lty=3, col='red')
  

tfit = treat(vfit, lfc=1)
dt = decideTests(tfit)

summary(dt)
```  
  

#Design including conditions and batch effect
```{r}
#Design
design=model.matrix(~0+conds+batch)
contr.matrix=makeContrasts(condsPrimary_Tumor-condsNormal, condsRecurrent_Tumor-condsNormal, levels = colnames(design))

quartz()
v=voom(dge, design, plot = T)

vfit=lmFit(v, design)
vfit=contrasts.fit(vfit, contrasts = contr.matrix)
efit=eBayes(vfit) #Does trend=1 need to be added here?
contrast2=efit

#PCA Plot
pca=prcomp(t(v$E))
quartz()
plot(pca$x, pch=c(21,24)[batch], bg=cond_colours[conds], cex=1, title(main = 'PCA Plot'))
text(pca$x,as.character(coldata$Batch),cex=0.3,pos=1)
text(pca$x,as.character(colnames(v$E)),cex=0.3,pos = 3)


#SA Plot
quartz()
plotSA(efit, main='Mean-Variance Trend')


#MDS Plot
quartz()
plotMDS(efit, pch = 19, col= cond_colours[conds], main='Multi Dimensional Scaling (MDS) Plot')
legend('bottomright', legend=levels(conds), col=cond_colours[unique(conds)], pch = 20)


#Summary of findings from comparisons
summary(decideTests(efit))


#Volcano Plots
quartz()
volcanoplot(efit, coef=1, names=as.character(gene_names[rownames(efit),1]), highlight = 50, main='Primary Tumour vs. Normal')
  abline(h=-log10(0.05), lty=3, col='red')
  abline(v=-1, lty=3, col='red')
  abline(v=1, lty=3, col='red')
  
quartz()
volcanoplot(efit, coef=2, names=as.character(gene_names[rownames(efit),1]), highlight = 50, main='Recurrent Tumour vs. Normal') 
  abline(h=-log10(0.05), lty=3, col='red')
  abline(v=-1, lty=3, col='red')
  abline(v=1, lty=3, col='red')

  
#Butterfly Plot - To compare how much the comparisons are similar/agree with eachother
plot(topTable(efit,coef=1,number=10000000000)$logFC,topTable(efit,coef=2,number=100000000000)$logFC)
  abline(h=0)
  abline(v=0)
  abline(a=0,b=1, lty=3, col='red')


#Further Analysis
tfit = treat(vfit, lfc=1)
dt = decideTests(tfit)
summary(dt)
```

#Compare Primary Tumour to TCGA Normals and to GTEX Normals Separately
```{r}
#Readjust Conditions Vectors to distinguish between GTEX and TCGA Normals - This will allow us to compare effect of comparing each individually
conds_origin=as.character(conds)
conds_origin[1:20]=rep("Normal_GTEX",20)
conds_origin[28:32]=rep("Normal_TCGA",5)
conds_origin=as.factor(conds_origin)

#Design changes to refelct changes in conds > conds_origin
design=model.matrix(~0+conds_origin+batch)
contr.matrix=makeContrasts(conds_originPrimary_Tumor-conds_originNormal_TCGA, conds_originPrimary_Tumor-conds_originNormal_GTEX, conds_originNormal_TCGA-conds_originNormal_GTEX, levels = colnames(design))

#Plot similarities between comparisons
plot(toptable(efit,coef=1,number=1000000000000000)$logFC,toptable(efit,coef=2,number=1000000000000000)$logFC)
abline(h=0)
abline(v=0)
abline(a=0,b=1, lty=1, col='red')
```


#Comparing normals with eachother
```{r}

design=model.matrix(~0+conds_origin+batch)
contr.matrix=makeContrasts(conds_originNormal_TCGA-conds_originNormal_GTEX, levels = colnames(design))


quartz()
volcanoplot(efit, coef=1, names=as.character(gene_names[rownames(efit),1]), highlight = 50, main='TCGA Normal vs. GTEX Normal')
  abline(h=-log10(0.05), lty=3, col='red')
  abline(v=-1, lty=3, col='red')
  abline(v=1, lty=3, col='red')
  
  
```


---------------------------------------------------------
---------------------------------------------------------


#Alternative Normalisation? - SVA
```{r}
#Need to create pdata-esque file for SVA
pdata=coldata

mod=model.matrix(~as.factor(conds), data = mergedcounts)
mod0=model.matrix(~1, data = pdata)
nsv=num.sv(rawcounts, mod, method = 'leek')
sva=sva(rawcounts, mod, mod0)

#Adjusting for Surrogate Variables
pValues=f.pvalue(test_counts,mod,mod0)
qValues=p.adjust(pValues, method='BH')

modSv = cbind(mod,sva$sv)
mod0Sv = cbind(mod0,sva$sv)
pValuesSv = f.pvalue(test_counts,modSv,mod0Sv)
qValuesSv = p.adjust(pValuesSv,method="BH")

#ComBat Function to adjust for batch effect between GTEX and TCGA - TO RETURN TO
batch
modcombat=model.matrix(~1, data = testmatrix)
combat_test=ComBat(dat = test_counts, batch = batch, mod=modcombat,par.prior=TRUE, prior.plots=FALSE)

#Simplified ComBat approach - TO RETURN TO 
modBatch=model.matrix(~as.factor(conds) + as.factor(batch), data = testmatrix)
mod0Batch=model.matrix(~as.factor(batch), data = testmatrix)

pValuesBatch = f.pvalue(test_counts,modBatch,mod0Batch)
qValuesBatch = p.adjust(pValuesBatch,method="BH")

#sva for sequencing
mod1=model.matrix(~conds)
modnull=cbind(mod1[,1])
svseq=svaseq(test_counts, mod1, modnull)$sv

plot(svseq, pch=19, col=cond_colours[conds])

Old TMM Normalisation Code

Type=factor(substring(colnames(mergedcounts),1))
Time=factor(substring(colnames(mergedcounts),1,))

dge=DGEList(counts=mergedcounts)

dge$samples$Type=conds
dge$samples$Batch=batch

dge=calcNormFactors(dge, method = 'TMM')

quartz()
boxplot(log2(dge$counts+1), las=2, col=cond_colours[conds], main='log2 counts', cex.axis=0.6)

quartz()
plotMDS(dge, pch = 19, col= cond_colours[conds], main='Multi Dimensional Scaling (MDS) Plot')
legend('bottomright', legend=levels(conds), col=cond_colours[unique(conds)], pch = 20)


design=model.matrix(~conds+batch, data = dge$samples)
rownames(design)=colnames(dge)

dge=estimateDisp(dge, design, robust=T)
dge$common.dispersion
quartz()
plotBCV(dge)

dge=estimateGLMCommonDisp(dge)
dge=estimateTagwiseDisp(dge)
fit = glmFit(dge, design, robust = T)

quartz()

fit2=glmQLFit(dge, design)
Compare GTEX normal to Primary Tumour (2), Recurrent Tumour (3), Solid 
qlf=glmQLFTest(fit2, coef = 2)
qlf=glmQLFTest(fit2, coef = 3)
qlf=glmQLFTest(fit2, coef = 4)
```

